### DISCUSSION TITLE
Control Rectangle Pro from scripts / commandline?

### DISCUSSION THREAD

**Discussion by NeighNeighNeigh** [rxhanson/RectanglePro-Community#98] [Q&A]
Hello,

Wondering if there's any way to control Rectangle through the command line or scripting? Basically I'd love to set up a whole bunch of custom shortcuts, but instead of needing keyboard shortcuts for all of them, I'd rather trigger them programatically if possible?

Cheers :)

**Comment by NeighNeighNeigh** [comment]
Maybe an elegant alternative might be through URI's. Add some kind of user defined handle in the custom entry, like so:
![Screen Shot 2022-09-20 at 20 26](https://user-images.githubusercontent.com/4795315/191236934-96e9d0aa-6ad7-4368-997e-a022be4fc093.png)

**Reply by rxhanson** [reply]
URI support will be in the next release of Rectangle, and after that goes out then I'm planning on rolling it into pro. When I do that, I'll most likely make it something that is based on the name of the commands/custom-shortcuts/application-groups, rather than something that is fully exposed to the user in the UI (at least for first cut).

**Reply by NeighNeighNeigh** [reply]
Beautiful! thanks Ryan that's fantastic to hear! Appreciate your great work.

**Reply by confluencepoint** [reply]
Was the feature added in version 2.6.3 (125)?
Does the URI still need to be registered in the system?

I get this error:
```
No application knows how to open URL 
rectangle://execute-action?name=left-half 
(Error Domain=NSOSStatusErrorDomain Code=-10814 "kLSApplicationNotFoundErr: E.g. no application 
claims the file" UserInfo={_LSLine=1489, _LSFunction=runEvaluator}).
```

**Reply by rxhanson** [reply]
Apologies for the lack of documentation, I’m traveling and had quickly pushed out the release before leaving. The URI for pro starts with `rectangle-pro` instead of `rectangle`.

**Reply by testcase** [reply]
Hi - Just wanted to clarify if this should be working. I am getting the following:
```
open -g "rectangle-pro://execute-action?name=left-half" 
No application knows how to open URL rectangle-pro://execute-action?name=left-half (Error Domain=NSOSStatusErrorDomain Code=-10814 "kLSApplicationNotFoundErr: E.g. no application claims the file" UserInfo={_LSLine=1489, _LSFunction=runEvaluator}).
```
running `lsregister -dump URLSchemeBinding` doesn't show rectangle-pro as being registered

**Reply by rxhanson** [reply]
What you have should work; it works for me. I have noticed macOS being a little wonky with registering things like this. Maybe a log out and log in will fix it, or uninstalling then reinstalling the app?

**Reply by testcase** [reply]
Great - yes a reinstall fixed the problem.

**Reply by SirStoke** [reply]
Hi! Is there a list of actions we're able to trigger via URIs? Specifically, I'm trying to trigger an application group layout

**Reply by rxhanson** [reply]
leftHalf, rightHalf, maximize, maximizeHeight, previousDisplay, nextDisplay, larger, smaller, bottomHalf, topHalf, center, bottomLeft, bottomRight, topLeft, topRight, restore, firstThird, firstTwoThirds, centerThird, lastTwoThirds, lastThird, moveLeft, moveRight, moveUp, moveDown, almostMaximize, fillLeft, fillRight, drag, resize, dial, reticle, manual, centerHalf, firstFourth, secondFourth, thirdFourth, lastFourth, topLeftSixth, topCenterSixth, topRightSixth, bottomLeftSixth, bottomCenterSixth, bottomRightSixth, firstSixth, lastSixth, fullscreen, close, minimize, quitApp, hideApp, winMod, winUp, winDown, winLeft, winRight, restoreToScreen, cascadeAll, cascadeApp, tile2x2, tile2x3, revealDesktopEdge, appNextDisplay, appPrevDisplay, appCollateLeft, appCollateRight, appLeftHalf, appRightHalf, specified, firstThreeFourths, lastThreeFourths, reverseAll, topLeftNinth, topCenterNinth, topRightNinth, middleLeftNinth, middleCenterNinth, middleRightNinth, bottomLeftNinth, bottomCenterNinth, bottomRightNinth, topLeftThird, topRightThird, bottomLeftThird, bottomRightThird, topLeftEighth, topCenterLeftEighth, topCenterRightEighth, topRightEighth, bottomLeftEighth, bottomCenterLeftEighth, bottomCenterRightEighth, bottomRightEighth, stashLeft, stashRight, stashUp, stashBottom, unstash, cycleStashed, toggleStashed, unstashAll, reflowTodo, hideAll, nextDisplayRatio, prevDisplayRatio, app, last, nudgeLeft, nudgeRight, nudgeUp, nudgeDown, snapTopLeft, snapTopRight, snapBottomLeft, snapBottomRight, centerTwoThirds, upperCenter, tidy

**Reply by rxhanson** [reply]
Application Groups & Custom Shortcuts will be in an upcoming release

**Reply by rxhanson** [reply]
App groups & custom shortcuts are now in 2.7.4:

Use `execute-layout` or `execute-custom`.

Examples:
`open -g "rectangle-pro://execute-layout?name=applayout1" `
`open -g "rectangle-pro://execute-custom?name=custom1" `

**Reply by b-ow** [reply]
Using 2.7.4 there is no URI field (see screenshot). What am I doing wrog?
<img width="835" alt="0DC665A9-8B9D-470B-97EC-BEEB4D321136" src="https://user-images.githubusercontent.com/122204226/211190029-2f057173-4bf3-444a-b6dd-b87081921370.png">

**Reply by NeighNeighNeigh** [reply]
@b-ow You shouldn't expect to see a URI field. rxhanson has implemented the feature without need for a field at all, instead you simply use the name of your custom shortcut in the URI.  

Example:
open -g "rectangle-pro://execute-custom?name=custom1"

**Reply by b-ow** [reply]
The list postet by rxhanson on Nov 30, 2022 "leftHalf, rightHalf, maximize,…" is at least partly incorrect. For example (using version 3.0) leftHalf does not work -left-half does; rightHalf doesn't work, right-half does; and so on. **Can you please provide a correct and complete list?** Thanks in advance!

**Reply by rxhanson** [reply]
I updated the Readme of this repo with the list (rather than camel case, it's hyphenated lowercase, as you found out).
https://github.com/rxhanson/RectanglePro-Community

**Comment by b-ow** [comment]
Worked. Thanks a lot.

BOW

> Am 08.01.2023 um 11:57 schrieb NeighNeighNeigh ***@***.***>:
> 
> open -g "rectangle-pro://execute-layout?name=applayout1 <rectangle-pro://execute-layout?name=applayout1>"
